### PR TITLE
Decaffeinate Tests

### LINKED ISSUE #7: Decaffeinate Tests
I noticed the following quote in https://github.com/ottypes/json1/issues/4#issuecomment-446064567, and thought I'd make an issue for it: 

> (And I'd like to port the tests away from coffeescript too!)

This issue can be closed when all the tests are ported to JavaScript.

I am thrilled to see this project at a steady slow burn. Really appreciate your long-term efforts on this!

### PR DESCRIPTION
**Pull request by curran**
Closes #7

### CODE REVIEW

**Comment by curran** [comment]
At this point, this is a rough first pass at the change, presenting the raw output from the automated tool https://github.com/decaffeinate/decaffeinate .

Looking through the resulting code, the added verbosity may benefit from a [Prettier](https://prettier.io/) pass (e.g. some lines are way too long now).

**Comment by curran** [comment]
I ran `prettier --single-quote ./*.js --write` in the `tests` directory, which resulted in a huge improvement.

**Comment by curran** [comment]
The remaing task here is to do a manual pass over all the generated JS files and clean them up. The generated `decaffeinate suggestions` comments may be helpful with this.

I'm willing to do this last leg of work as well, but I'd like to hear from you @josephg on the general approach of this PR. What do you think? Would this be a welcome change? Thanks!

**Comment by josephg** [comment]
Yep this looks great! I do appreciate this change, and like you I'm curious why the tests are failing. I'd also like it if decaffinate didn't add semicolons at the end of all my lines (my poor hipster JS style conventions...). Once you sort that out I'll merge the change.

A better way to verify immutability would be to recursively call `Object.freeze` on all inputs to all the functions in all tests. The nice thing about that approach is that if any tests inappropriately modify data, the stack trace will point to the exact source of the error. If we do that, we probably shouldn't need special immutability test cases at all.

**Comment by curran** [comment]
Awesome! Thanks for reviewing. I will work on the points you mentioned.

I can tell Prettier not to include semicolons, so that change at least will be automatic.

Changing the way that the immutability tests work at first glance feels out of scope for this PR, but maybe it's necessary now if the decaffeination changed something subtle/fundamental with those particular tests.

**Comment by josephg** [comment]
Yes I agree it’s out of scope. And I’d like to know what is causing the problem anyway, on the off chance it’s independently something worth fixing.

**Comment by curran** [comment]
For the record, this is the prettier command I used:

`prettier --single-quote --no-semi ./*.js --write`

**Comment by curran** [comment]
Ok I found the cause of the error.

The test uses a `require` call to pull in a `.coffee` file. In `test/genOp.js`:

```
194         const genString = require('ot-text/test/genOp') // Tries loading genOp.coffee
```

This should work if we bring back the `coffeescript/register` in mocha config.

It would not be a total decaffeination, but it's pretty close.

**Comment by curran** [comment]
Woohoo CI passes! Onto the manual cleanup.

**Comment by curran** [comment]
@josephg Ready for final review. I did a manual pass through the code and removed unnecessary stuff added by `decaffeinate` (mostly unnecessary `return` statements and IIFEs).

I would like to bring your attention to this line in the original test, which doesn't seem to make sense https://github.com/ottypes/json1/blob/d8c998d3ce9673591b155d6ea4e57dfb780e19db/test/test.coffee#L2003 .

```
      expect: [ [ 'a', { r: true }, 0, { p: 0 } ], [ 'b', { d: 0 } ] ]
```

I've added a `TODO` flag for you on the code generated from that line, right here https://github.com/ottypes/json1/pull/8/files#diff-c1129c8b045390789fa8ff62f2c6b4a9R2781

```
      // TODO have a look at this - doesn't seem right to just define an object
      {
        expect: [['a', { r: true }, 0, { p: 0 }], ['b', { d: 0 }]]
      }
```

Thanks for accepting my contribution!

File: test/genOp.js [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    w.ascend()
+    w.reset()
+    path.pop()
+  }
+  key = randomNewKey(parent)
+  return [path, parent, key]
+}
+
+const set = (container, key, value) =>
+  value === undefined
```
Comment (curran) [inline test/genOp.js]: By the way, I manually changed a coupld of these `if` `else` sections to use these ternary operators instead. Happy to switch them back if you'd prefer.

**Comment by josephg** [comment]
> I would like to bring your attention to this line in the original test, which doesn't seem to make sense
> 
> [json1/test/test.coffee](https://github.com/ottypes/json1/blob/d8c998d3ce9673591b155d6ea4e57dfb780e19db/test/test.coffee#L2003)
> Line 2003 in [d8c998d](/ottypes/json1/commit/d8c998d3ce9673591b155d6ea4e57dfb780e19db)
>  expect: [ [ 'a', { r: true }, 0, { p: 0 } ], [ 'b', { d: 0 } ] ] 
> .
> ```
>       expect: [ [ 'a', { r: true }, 0, { p: 0 } ], [ 'b', { d: 0 } ] ]
> ```

Yeah I have no idea what thats doing there. Its a remnant of some test case, no doubt. Its safe to delete.

I'm not entirely happy that coffeescript is still a dependancy after this change - though I assume we'll be able to remove it with a bit more refactoring.

But yeah, great stuff! I'm going to merge this.

**Comment by curran** [comment]
Hooray!

**Pull request merged**